src/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,179 @@
+        LOG(VISUALIZE) << *createDebugShapes(debug_shapes);
+    }
+
+    // Generate sample passes across the field for cost visualization
+    if (passing_config_.cost_vis_config().generate_sample_passes())
+    {
+        samplePassesForVisualization(world, passing_config_, best_pass.pass);
+    }
+
+    return best_pass;
+}
+
+std::map<RobotId, std::vector<Point>> PassGenerator::sampleReceivingPositionsPerRobot(
+    const World& world, const std::vector<RobotId>& robots_to_ignore)
+{
+    std::map<RobotId, std::vector<Point>> receiving_positions_map;
+
+    const double min_sampling_std_dev =
+        passing_config_.pass_gen_min_rand_sample_std_dev_meters();
+    const double sampling_std_dev_vel_multiplier =
+        passing_config_.pass_gen_rand_sample_std_dev_robot_vel_multiplier();
+    const double sampling_center_vel_multiplier =
+        passing_config_.pass_gen_rand_sample_center_robot_vel_multiplier();
+
+    for (const Robot& robot : world.friendlyTeam().getAllRobots())
+    {
+        // Ignore robots in the ignore list
+        if (std::find(robots_to_ignore.begin(), robots_to_ignore.end(), robot.id()) !=
+            robots_to_ignore.end())
+        {
+            continue;
+        }
+
+        // Add the robot's current position to the list of sampled passes
+        auto robot_position = robot.position();
+        receiving_positions_map.insert({robot.id(), {robot_position}});
+
+        // Add the best pass from the previous iteration to the list of sampled passes
+        if (previous_best_receiving_positions_.find(robot.id()) !=
+            previous_best_receiving_positions_.end())
+        {
+            receiving_positions_map[robot.id()].push_back(
+                previous_best_receiving_positions_[robot.id()]);
+        }
+
+        // Sample random receiving positions using a normal distribution around the
+        // robot's future position with a standard deviation that scales with the robot's
+        // velocity.
+        Point sampling_center =
+            robot_position + (robot.velocity() * sampling_center_vel_multiplier);
+        double std_dev = min_sampling_std_dev +
+                         (robot.velocity().length() * sampling_std_dev_vel_multiplier);
+        std::normal_distribution x_normal_distribution{sampling_center.x(), std_dev};
+        std::normal_distribution y_normal_distribution{sampling_center.y(), std_dev};
+
+        for (unsigned int i = 0; i < passing_config_.pass_gen_num_samples_per_robot();
+             i++)
+        {
+            auto point = Point(x_normal_distribution(random_num_gen_),
+                               y_normal_distribution(random_num_gen_));
+            // Only consider points within the playing area
+            if (contains(world.field().fieldLines(), point))
+            {
+                receiving_positions_map[robot.id()].push_back(point);
+            }
+        }
+    }
+
+    return receiving_positions_map;
```

Comment:
So something like `return std::move(receiving_positions_map)` ? We don't really want to make a copy?

Reply:
Nvm this is dangerous